File: Data-Structures-And-Algorithms/Map/leetCode-3-lengthOfLongestSubstring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        ans = 0
        d = {}
        for k, t in enumerate(s):
            if t in d:
                left = max(d[t]+1, left)
            d[t] = k
            ans = max(ans, k-left+1)

        return ans

File: Data-Structures-And-Algorithms/Map/leetCode-3-lengthOfLongestSubstring.go
func lengthOfLongestSubstring(s string) int {
	left, ans := 0, 0
	m := map[byte]int{}
	for i := 0; i < len(s); i++ {
		if _, ok := m[s[i]]; ok { // 这里注意：判断map中是否存在当前key，用if这种方法
			left = max(m[s[i]]+1, left)
		}
		m[s[i]] = i
		ans = max(ans, i-left+1)
	}
	return ans
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
